Improvements and documentation

calvinlfer opened the pull request:
- Add documentation 
- Support ECR being in the same account as Jenkins via a CloudFormation template
- Allow conditional setup of a Kinesis subscription filter
- Allow specifying a Docker image to use (default to `loyaltyone/jenkins`) for the Jenkins master
- Group CloudFormation parameters
- Feeding in additional parameters to Jenkins master in Task Definition to speed up slave provisioning (see [here](https://github.com/jenkinsci/amazon-ecs-plugin/pull/48#issuecomment-374554410) for more information)

calvinlfer committed:
Document the purpose of SlaveRoles.yml

calvinlfer committed:
Merge pull request #1 from ukayani/master

fix hardcoding of accountID

calvinlfer committed:
Document the purpose of *SlaveRoles.yml and create one that supports
ECR residing in the same AWS account as Jenkins

ukayani commented on LocalSlaveRoles.yml:
```diff
@@ -0,0 +1,54 @@
+          - Effect: Allow
+            Action:
+              - ecr:*
+            Resource: !Ref 'ECRRepoArn'
```
I thought you were planning on setting this to a wildcard with the current accountId as part of the arn

calvinlfer replied:
I can definitely do that instead (`arn:aws:ecr:${AWS::Region}:${AWS::AccountId}:repository/*`), its just that if you want to host multiple ECR repositories and multiple Jenkins clusters in a sandboxed fashion then this using the wildcard would mess things up, if that's something you would like to support (i'm not really sure whether this concern is even valid). 

Otherwise you just expect to have one Jenkins cluster per account. 

What do you think?

calvinlfer replied:
Nice catch. I have incorporated this now.

calvinlfer committed:
Perform conditional creation of the Kinesis Subscription Filter

calvinlfer committed:
Simplify template
- this IAM role will allow any Jenkins slave tasks that use it to be able to push Docker images to any ECR repository residing in the same account

calvinlfer committed:
I forgot to add ecr:GetAuthorizationToken for the case where ECR
resides in the same account as Jenkins.

You need this permission to get credentials to push to repositories

For example:
```bash
# Login to private container repository
eval $(aws ecr get-login --no-include-email)
```

calvinlfer committed:
- Allow specifying a Docker image to use for the Jenkins master
- Group CloudFormation parameters

calvinlfer committed:
Freckle specific changes

calvinlfer committed:
Add additional startup parameters to speed up provisioning and running
of Jenkins agents for the ECS plugin

@see https://github.com/jenkinsci/amazon-ecs-plugin/pull/48#issuecomment-374554410
